Lib/multiprocessing/pool.py

```diff
@@ -662,13 +662,14 @@ def __exit__(self, exc_type, exc_val, exc_tb):
 
 class ApplyResult(object):
 
-    def __init__(self, cache, callback, error_callback):
+    def __init__(self, pool, callback, error_callback, cache=None):
+        self._pool = pool
```

Comment:
You wrote "This PR fixes a regression introduced in PR #8450" which comes from https://bugs.python.org/issue34172. This issue is a reference cycle fix if I understand correctly, but here you create a new strong reference to the pool. I see a risk of creating *new* reference cycles. No?

Before https://bugs.python.org/issue34172 fix, ApplyResult didn't hold a reference to the pool: you add a **new** reference.

Maybe it's fine, but I'm just concerned that this PR might reintroduce a bug similar but different than  https://bugs.python.org/issue34172

I don't understand why you need to hold a strong reference to the pool.

Reply:
We need to hold a reference to the pool because the regression is that the pool may be destroyed while one of its iterators is still alive if no references are left to the pool. In this case, iterating over the iterator of the pool hangs. The solution links the lifetime of the pool to the lifetime of the iterator (basically: keeping the pool alive while the iterator is alive). When the iterator is exhausted, the references to the pool are cleaned. 

The issue in #8450 is different, is a reference cycle between the Pool object, and the `self._worker_handlerp  Thread object, which is more fundamental.

In this case, the cycle needs to exist while the iterator is alive: all objects that depend on the pool need to make sure that the pool is alive while they are still in use (a non-consumed iterator...etc), otherwise hangs or inconsistent state can happen.

Reply:
More long explanation available here:

https://bugs.python.org/msg330996